### ISSUE TITLE
regexp segfault

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6433]
Migrated from [rt.perl.org#21922](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=21922) (status was 'resolved')

Searchable as RT21922$

**Comment by p5pRT** [comment]
### From powerman@sky.net.ua

#### Created by powerman@sky.net.ua




$ perl \-e '$\_ = " " x 50000; /\(?&#8203;:  |\)\*/'
Segmentation fault
$ perl \-e '$\_ = " " x 50000; /\(?&#8203;:  |a\)\*/'
Segmentation fault
$ perl \-e '$\_ = " " x 50000; /\(?&#8203;: |aa\)\*/'
Segmentation fault
$ perl \-e '$\_ = " " x 50000; /\(?&#8203;:aa|\[^a\]\)\*/'
Segmentation fault
$ perl \-e '$\_ = " " x 50000; /\(?&#8203;:ab|\[^c\]\)\*/'
Segmentation fault

There a lot of other regexps like these which also die by segfault
if string is long enough \(usually string should be >4100 bytes\,
but critical string size depend on environment \[perl binary?\]\)\.

I've tested this bug on perl 5\.8\.0 and 5\.6\.1\.

I've extracted these simple examples from very very big regexp\.
And I don't really understand what and why happens\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=critical

Site configuration information for perl v5.8.0:

Configured by root at Tue Feb 11 17:22:15 EET 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.19, archname=i686-linux-thread-multi
    uname='linux home.power 2.4.19 #1 ×ÓË Á×Ç 18 00:22:21 eest 2002 i686 unknown '
    config_args='-Dprefix=/usr -Doptimize= -O3 -march=athlon -mcpu=athlon  -d -e -s -Dinstallprefix=/usr -Dusethreads'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize=' -O3 -march=athlon -mcpu=athlon ',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing'
    ccversion='', gccversion='3.0', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -ldbm -ldb -ldl -lm -lpthread -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lcrypt -lutil
    libc=/lib/libc-2.2.5.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/lib/perl5/5.8.0/i686-linux-thread-multi
    /usr/lib/perl5/5.8.0
    /usr/lib/perl5/site_perl/5.8.0/i686-linux-thread-multi
    /usr/lib/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/home/powerman
    LANG=ru_RU.koi8r
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/powerman/bin:/bin:/usr/bin:/sbin:/usr/sbin:/usr/X11R6/bin:/usr/local/bin:/usr/local/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From enache@rdslink.ro

On Thu\, Apr 10\, 2003 at 01&#8203;:44&#8203;:50AM \-0000\, powerman@&#8203;sky\.net\.ua \(via RT\) wrote&#8203;:
> $ perl \-e '$\_ = " " x 50000; /\(?&#8203;:  |\)\*/'
> Segmentation fault
\[snip\]

Hard stack overflow \- because the regexp engine is recursive\.
Try increasing your stack size limit \-
or designing a new non\-recursive regexp engine :\)

$ ulimit \-s
8192
$ perl \-e '$\_ = " " x 50000; /\(?&#8203;:  |\)\*/'
Segmentation fault
$ ulimit \-s 24000
$ perl \-e '$\_ = " " x 50000; /\(?&#8203;:  |\)\*/'
$

\('limit stacksize 24000' if you're using csh\)

Regards
Adi

**Comment by p5pRT** [comment]
### From schubiger@cpan.org

> \[enache \- Wed Apr 09 19&#8203;:40&#8203;:24 2003\]&#8203;:
>
> Hard stack overflow \- because the regexp engine is recursive\.
> Try increasing your stack size limit \-
> or designing a new non\-recursive regexp engine :\)
> 
> $ ulimit \-s
> 8192
> $ perl \-e '$\_ = " " x 50000; /\(?&#8203;:  |\)\*/'
> Segmentation fault
> $ ulimit \-s 24000
> $ perl \-e '$\_ = " " x 50000; /\(?&#8203;:  |\)\*/'
> $
> 
> \('limit stacksize 24000' if you're using csh\)
> 
> Regards
> Adi

Answer and examples provided; can this ticket be put to rest?

**Comment by p5pRT** [comment]
### From @smpeters

> \[powerman@&#8203;sky\.net\.ua \- Wed Apr 09 18&#8203;:44&#8203;:49 2003\]&#8203;:
> 
> 
> This is a bug report for perl from powerman@&#8203;sky\.net\.ua\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> 
> $ perl \-e '$\_ = " " x 50000; /\(?&#8203;:  |\)\*/'
> Segmentation fault
> $ perl \-e '$\_ = " " x 50000; /\(?&#8203;:  |a\)\*/'
> Segmentation fault
> $ perl \-e '$\_ = " " x 50000; /\(?&#8203;: |aa\)\*/'
> Segmentation fault
> $ perl \-e '$\_ = " " x 50000; /\(?&#8203;:aa|\[^a\]\)\*/'
> Segmentation fault
> $ perl \-e '$\_ = " " x 50000; /\(?&#8203;:ab|\[^c\]\)\*/'
> Segmentation fault
> 
> There a lot of other regexps like these which also die by segfault
> if string is long enough \(usually string should be >4100 bytes\,
> but critical string size depend on environment \[perl binary?\]\)\.
> 
> I've tested this bug on perl 5\.8\.0 and 5\.6\.1\.
> 
> I've extracted these simple examples from very very big regexp\.
> And I don't really understand what and why happens\.
> 
> 

This has been fixed with change \#27598\.

steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ perl \-e '$\_ = " " x 50000; /\(?&#8203;:  |\)\*/'
Segmentation fault
steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ perl \-e '$\_ = " " x 50000; /\(?&#8203;:  |a\)\*/'
Segmentation fault
steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ perl \-e '$\_ = " " x 50000; /\(?&#8203;: |aa\)\*/'
Segmentation fault
steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ perl \-e '$\_ = " " x 50000; /\(?&#8203;:aa|\[^a\]\)\*/'
Segmentation fault
steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ perl \-e '$\_ = " " x 50000; /\(?&#8203;:ab|\[^c\]\)\*/'
Segmentation fault
steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ \./perl \-e '$\_ = " " x 50000; /\(?&#8203;:  |\)\*/'
steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ \./perl \-e '$\_ = " " x 50000; /\(?&#8203;:  |a\)\*/'
steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ \./perl \-e '$\_ = " " x 50000; /\(?&#8203;: |aa\)\*/'
steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ \./perl \-e '$\_ = " " x 50000; /\(?&#8203;:aa|\[^a\]\)\*/'
steve@&#8203;kirk&#8203;:\~/smoke/perl\-current$ \./perl \-e '$\_ = " " x 50000; /\(?&#8203;:ab|\[^c\]\)\*/'

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
